File: app/src/main/java/org/rivchain/cuplink/util/StatsReportUtil.java
package org.rivchain.cuplink.util;

import org.webrtc.RTCStats;
import org.webrtc.RTCStatsReport;

import java.math.BigInteger;
import java.util.Map;

/**
 * @Author Vadym Vikulin
 * @Date 2021/10/10
 * @desc
 */
public class StatsReportUtil {

    private final static int STATS_INTERVAL_MS = 5000;
    private BigInteger lastBytesReceivedVideo = BigInteger.valueOf(0);
    private BigInteger lastBytesSentVideo = BigInteger.valueOf(0);
    private BigInteger lastBytesReceivedAudio = BigInteger.valueOf(0);
    private BigInteger lastBytesSentAudio = BigInteger.valueOf(0);
    private Long lastFrameDecoded = Long.valueOf(0);

    public String getStatsReport(RTCStatsReport report) {
        if (report == null) return "";
        String codecIdVideo = null;
        String codecIdAudio = null;
        String codecVideo = "";
        String codecAudio = "";
        long receivedBytesSRVideo = 0;
        long sentBytesSRVideo = 0;
        long receivedBytesSRAudio = 0;
        long sentBytesSRAudio = 0;
        long width = 0, height = 0;
        long frameRate = 0;
        for (RTCStats stats : report.getStatsMap().values()) {
            if (stats.getType().equals("inbound-rtp")) {
                Map<String, Object> members = stats.getMembers();
                if (members.get("mediaType").equals("video")) {
                    codecIdVideo = (String) members.get("codecId");
                    BigInteger bytes = (BigInteger) members.get("bytesReceived");
                    receivedBytesSRVideo = bytes.longValue() - lastBytesReceivedVideo.longValue();
                    lastBytesReceivedVideo = bytes;
                    long currentFrame = (long) members.get("framesDecoded");
                    long lastFrame = lastFrameDecoded;
                    frameRate = (currentFrame - lastFrame) * 1000
                            / STATS_INTERVAL_MS;
                    lastFrameDecoded = currentFrame;
                }
                if (members.get("mediaType").equals("audio")) {
                    codecIdAudio = (String) members.get("codecId");
                    BigInteger bytes = (BigInteger) members.get("bytesReceived");
                    receivedBytesSRAudio = bytes.longValue() - lastBytesReceivedAudio.longValue();
                    lastBytesReceivedAudio = bytes;
                }
            }
            if (stats.getType().equals("outbound-rtp")) {
                Map<String, Object> members = stats.getMembers();
                if (members.get("mediaType").equals("video")) {
                    BigInteger bytes = (BigInteger) members.get("bytesSent");
                    sentBytesSRVideo = bytes.longValue() - lastBytesSentVideo.longValue();
                    lastBytesSentVideo = bytes;
                }
                if (members.get("mediaType").equals("audio")) {
                    BigInteger bytes = (BigInteger) members.get("bytesSent");
                    sentBytesSRAudio = bytes.longValue() - lastBytesSentAudio.longValue();
                    lastBytesSentAudio = bytes;
                }
            }
            if (stats.getType().equals("track")) {
                Map<String, Object> members = stats.getMembers();
                if (members.get("kind").equals("video")) {
                    width = members.get("frameWidth") == null ? 0 : (long) members.get(
                            "frameWidth");
                    height = members.get("frameHeight") == null ? 0 : (long) members.get(
                            "frameHeight");
                }
            }
        }
        if (codecIdVideo != null) {
            codecVideo = (String) report.getStatsMap().get(codecIdVideo).getMembers().get("mimeType");
        }
        if (codecIdAudio != null) {
            codecAudio = (String) report.getStatsMap().get(codecIdAudio).getMembers().get("mimeType");
        }

        final String statsReport = "Codecs: " + codecVideo + " " + codecAudio
                + "\nResolution: " + width + "x" + height
                + "\nBitrate ⎚ ↓: " + receivedBytesSRVideo * 8 / STATS_INTERVAL_MS + "kbps"
                + "\nBitrate ⎚ ↑: " + sentBytesSRVideo * 8 / STATS_INTERVAL_MS + "kbps"
                + "\nBitrate \uD83D\uDD0A ↓: " + receivedBytesSRAudio * 8 / STATS_INTERVAL_MS + "kbps"
                + "\nBitrate \uD83D\uDD0A ↑: " + sentBytesSRAudio * 8 / STATS_INTERVAL_MS + "kbps"
                + "\nFrameRate: " + frameRate;
        return statsReport;
    }
}

File: app/src/main/java/d/d/meshenger/call/StatsReportUtil.kt
package d.d.meshenger.call

import org.webrtc.RTCStatsReport
import java.math.BigInteger

/**
 * @Author Vadym Vikulin
 * @Date 2021/10/10
 * @desc
 */
class StatsReportUtil {
    private var lastBytesReceivedVideo = BigInteger.valueOf(0)
    private var lastBytesSentVideo = BigInteger.valueOf(0)
    private var lastBytesReceivedAudio = BigInteger.valueOf(0)
    private var lastBytesSentAudio = BigInteger.valueOf(0)
    private var lastFrameDecoded = java.lang.Long.valueOf(0)
    fun getStatsReport(report: RTCStatsReport?): String {
        if (report == null) return ""
        var codecIdVideo: String? = null
        var codecIdAudio: String? = null
        var codecVideo: String? = ""
        var codecAudio: String? = ""
        var receivedBytesSRVideo: Long = 0
        var sentBytesSRVideo: Long = 0
        var receivedBytesSRAudio: Long = 0
        var sentBytesSRAudio: Long = 0
        var width: Long = 0
        var height: Long = 0
        var frameRate: Long = 0
        for (stats in report.statsMap.values) {
            if (stats.type == "inbound-rtp") {
                val members = stats.members
                if (members["mediaType"] == "video") {
                    codecIdVideo = members["codecId"] as String?
                    val bytes = members["bytesReceived"] as BigInteger?
                    receivedBytesSRVideo = bytes!!.toLong() - lastBytesReceivedVideo.toLong()
                    lastBytesReceivedVideo = bytes
                    val currentFrame = members["framesDecoded"] as Long
                    val lastFrame = lastFrameDecoded
                    frameRate = ((currentFrame - lastFrame) * 1000
                            / STATS_INTERVAL_MS)
                    lastFrameDecoded = currentFrame
                }
                if (members["mediaType"] == "audio") {
                    codecIdAudio = members["codecId"] as String?
                    val bytes = members["bytesReceived"] as BigInteger?
                    receivedBytesSRAudio = bytes!!.toLong() - lastBytesReceivedAudio.toLong()
                    lastBytesReceivedAudio = bytes
                }
            }
            if (stats.type == "outbound-rtp") {
                val members = stats.members
                if (members["mediaType"] == "video") {
                    val bytes = members["bytesSent"] as BigInteger?
                    sentBytesSRVideo = bytes!!.toLong() - lastBytesSentVideo.toLong()
                    lastBytesSentVideo = bytes
                }
                if (members["mediaType"] == "audio") {
                    val bytes = members["bytesSent"] as BigInteger?
                    sentBytesSRAudio = bytes!!.toLong() - lastBytesSentAudio.toLong()
                    lastBytesSentAudio = bytes
                }
            }
            if (stats.type == "track") {
                val members = stats.members
                if (members["kind"] == "video") {
                    width = if (members["frameWidth"] == null) 0 else members["frameWidth"] as Long
                    height =
                        if (members["frameHeight"] == null) 0 else members["frameHeight"] as Long
                }
            }
        }
        if (codecIdVideo != null) {
            codecVideo = report.statsMap[codecIdVideo]!!.members["mimeType"] as String?
        }
        if (codecIdAudio != null) {
            codecAudio = report.statsMap[codecIdAudio]!!.members["mimeType"] as String?
        }
        return """Codecs: $codecVideo $codecAudio
                Resolution: ${width}x$height
                Bitrate ⎚ ↓: ${receivedBytesSRVideo * 8 / STATS_INTERVAL_MS}kbps
                Bitrate ⎚ ↑: ${sentBytesSRVideo * 8 / STATS_INTERVAL_MS}kbps
                Bitrate 🔊 ↓: ${receivedBytesSRAudio * 8 / STATS_INTERVAL_MS}kbps
                Bitrate 🔊 ↑: ${sentBytesSRAudio * 8 / STATS_INTERVAL_MS}kbps
                FrameRate: $frameRate"""
    }

    companion object {
        private const val STATS_INTERVAL_MS = 5000
    }
}